frame/support/src/traits/bounded_encoded_length.rs

```diff
@@ -0,0 +1,95 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use codec::{Compact, Encode};
+use impl_trait_for_tuples::impl_for_tuples;
+use sp_std::{mem, marker::PhantomData};
+
+/// Items implementing `BoundedEncodedLen` have a statically known maximum encoded size.
+///
+/// Some containers, such as `BoundedVec`, have enforced size limits and this trait
+/// can be implemented accurately. Other containers, such as `StorageMap`, do not have enforced size
+/// limits. For those containers, it is necessary to make a documented assumption about the maximum
+/// usage, and compute the max encoded length based on that assumption.
+pub trait BoundedEncodedLen: Encode {
+	/// Upper bound, in bytes, of the maximum encoded size of this item.
+	fn max_encoded_len() -> usize;
+}
+
+macro_rules! impl_primitives {
+	( $($t:ty),+ ) => {
+		$(
+			impl BoundedEncodedLen for $t {
+				fn max_encoded_len() -> usize {
+					mem::size_of::<$t>()
+				}
+			}
+		)+
+	};
+}
+
+impl_primitives!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
+
+impl<T> BoundedEncodedLen for Compact<T>
+where
+	Compact<T>: Encode,
+	T: BoundedEncodedLen,
+{
+	// The compact encoding of a type usually requires fewer bytes, but can occupy 1 additional
+	// byte in the worst case.
```

Comment:
is it true ? Compact<u16>(2^14) takes 4 bytes to be encoded in scale codec, whereas u16 it takes only 2.

I think you need to derive for each primitive integer types as it depends it has different worst case

Reply:
You're right. Good catch! I hadn't realized our encoding worked like that. Fix incoming.

Reply:
`u16` is the only type which takes other than one extra byte. I wish I knew why the codec is written like that; it seems suboptimal.

Reply:
I'm not sure it is suboptimal, any codec is a tradeoff, if we have a different encoding allowing u16 to be encoded in 3 bytes, we then have some other value to be encoded in more bytes than currently.